pkg/cluster/cert.go

```diff
@@ -168,6 +185,7 @@ func GetSecret(ctx context.Context, c client.Client, namespace, name string) (*v
 	return secret, nil
 }
 
+// copySecretToNamespace copies secret to given namespace with exact same data and type.
 func copySecretToNamespace(ctx context.Context, c client.Client, secret *v1.Secret, newSecretName, namespace string) error {
```

Comment:
Will that work:
```golang
// syncSecretType return true if existingSecret is valid
func syncSecretType(ctx context.Context, c client.Client, secret, existingSecret *v1.Secret) (bool, error) {
	name := secret.ObjectMeta.Name
	ns   := secret.ObjectMeta.Name

	err := c.Get(ctx, client.ObjectKey{Name: name, Namespace: ns}, existingSecret)

	switch {
	case err == nil:
		if existingSecret.Type == secret.Type {
			return true, nil
		}
		// Secret exists but with a different type, delete and recreate it
		if err := c.Delete(ctx, existingSecret); err != nil {
			return false, fmt.Errorf("failed deleting existing secret: %w", err)
		}
		if err := c.Create(ctx, secret); client.IgnoreAlreadyExists(err) != nil {
			return false, fmt.Errorf("failed creating certificate secret: %w", err)
		}

	case apierrors.IsNotFound(err):
		// Secret does not exist, create it
		if err := c.Create(ctx, secret); err != nil {
			return false, fmt.Errorf("failed creating certificate secret: %w", err)
		}
	default:
		return false, fmt.Errorf("failed getting certificate secret: %w", err)
	}

	// never reached
	return false, nil
}

func CreateSelfSignedCertificate(ctx context.Context, c client.Client, secretName, domain, namespace string, metaOptions ...MetaOptions) error {
	certSecret, err := GenerateSelfSignedCertificateAsSecret(secretName, domain, namespace)
	if err != nil {
		return fmt.Errorf("failed generating self-signed certificate: %w", err)
	}

	if err := ApplyMetaOptions(certSecret, metaOptions...); err != nil {
		return err
	}

	existingSecret := &v1.Secret{}
	_, err = syncSecretType(ctx, c, certSecret, existingSecret)
	return err
}

func copySecretToNamespace(ctx context.Context, c client.Client, secret *v1.Secret, newSecretName, namespace string) error {
	newSecret := &v1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name:      newSecretName,
			Namespace: namespace,
		},
		Data: secret.Data,
		Type: secret.Type,
	}

	existingSecret := &v1.Secret{}
	ok, err := syncSecretType(ctx, c, newSecret, existingSecret)
	if err != nil {
		return err
	}
	
	if ok && isSecretOutdated(existingSecret.Data, newSecret.Data) {
		if err = c.Update(ctx, newSecret); err != nil { // update data if found with same type but outdated content
			return fmt.Errorf("failed to update secret: %w", err)
		}
	}

	return nil
}
```
? Completely untested.

Reply:
i did some update for the delete-create into a function, 
do you think it is clear enough to maintain this part?

Reply:
It anyway introduces duplication of the same logic "if it does not exist, create it, else recreate if it has wrong type" into two places.